engine, resources: Adds symbolic mode parsing to file resource

dgbuckley opened the pull request:
Adds symbolic mode parsing functionality to the "util" package for parsing in the
file resource.

purpleidea commented:
@dgbuckley Thanks for the patch!

As a funny coincidence, a mentee of the project was also working on this issue as a learning exercise, however this version is much closer to what we're looking for.

There are a few review items which I hope you wouldn't mind fixing, but I was wondering if you would be okay waiting a short while to give my mentee a chance to practice reviewing this first.

@dbakong (currently learning golang) will review as a learning exercise. Please ignore those comments. After his are submitted, I'll do a final review.

Hope that's okay, and thank you for your patch!

purpleidea commented on util/sym-mode_test.go:
```diff
@@ -0,0 +1,58 @@
+		input  string
+		expect os.FileMode
+	}{
+		{"ug=rwx", 0774 | os.ModeSetgid},
```
Or maybe that should be an error?

dgbuckley commented:
@purpleidea that sounds great. No rush on this PR, I'm just excited to be contributing to this awesome project. Also sorry about the overlap of work; I'll do better to watch PRs moving forward.

purpleidea commented on engine/resources/file.go:
```diff
@@ -137,9 +137,19 @@ func (obj *FileRes) isDir() bool {
 func (obj *FileRes) mode() (os.FileMode, error) {
-	m, err := strconv.ParseInt(obj.Mode, 8, 32)
+	n, err := strconv.ParseInt(obj.Mode, 8, 32)
+	m := os.FileMode(n)
```
Having this between the `xxx, err := ...` and the `if err != nil` is awkward, might be better to inline the code and exit early if `err == nil`.

purpleidea commented on engine/resources/file.go:
```diff
@@ -137,9 +137,19 @@ func (obj *FileRes) isDir() bool {
-		return os.FileMode(0), errwrap.Wrapf(err, "mode should be an octal number (%s)", obj.Mode)
+		// Try parsing symbolic
+		modes := strings.Split(obj.Mode, ",")
+		m = os.FileMode(0)
```
If this is the initial condition, don't we want this to be the mode of the current file since we might +/- on it?

dgbuckley replied:
If there is an error getting the mode from the file should it be raised then or just set to 0 and have that handled elsewhere?

purpleidea replied:
The first one. So if when trying to read the file mode that errors, then we propagate that error up and the engine handles it.

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
```
use our util errwrap package instead please

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+	"github.com/pkg/errors"
+)
+
+const (
```
public const block needs a comment for golint to pass IIRC.

purpleidea commented on engine/resources/file.go:
```diff
@@ -137,9 +137,19 @@ func (obj *FileRes) isDir() bool {
 	if err != nil {
-		return os.FileMode(0), errwrap.Wrapf(err, "mode should be an octal number (%s)", obj.Mode)
+		// Try parsing symbolic
+		modes := strings.Split(obj.Mode, ",")
```
As per @dbakong's comment, I think we should track duplicates so that we don't allow `u=rw,u=r` or `u+rw,u-r` sorts of things. As a result of adding that logic, perhaps this should be in a separate helper function?

dgbuckley replied:
My initial interpretation of symbolic mode is that it should be idempotent and it shouldn't matter if you gave `u=x,u=r,u=rw` you would just get `u=rw`. Thinking about it further that may allow for some confusing inputs. Should repeated subjects (ex uug=rw) be an error as well?

purpleidea replied:
Honestly, unless the `chmod` "standard" allows such weird things (please have a look) then we shouldn't allow them at all. We want to be less permissive on what we accept so that accidentally incorrect input here causes an error instead of silently allowing wrong behaviour.

As a slight warning: I was planning to write this patch (but you're welcome to include it if you like) but while it's great that you support the `+` and `-`, I was thinking of having those be enabled by a bool const. That way, we can leave support for those off by default. My reasoning: I'm contemplating whether they're really necessary here or not. But I like having the possibility to flip it on if we find a good reason. Cheers!

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+		return modeSubtracted(mode, from)
+	}
+
+	return from, errors.Errorf("%s is not a valid symbolic mode", mode)
```
fmt.Errorf, and the return value should be the zero or nil.
Same in other places too.

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+// Ex. ug=rw
+func SymbolicMode(mode string, from os.FileMode) (os.FileMode, error) {
+	// parse the who
+	// var i int
```
dead code?

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+func SymbolicMode(mode string, from os.FileMode) (os.FileMode, error) {
+	// parse the who
+	// var i int
+	if strings.Contains(mode, "=") && !strings.ContainsAny(mode, "+-") {
```
Why isn't this `mode == "="` ?

dgbuckley replied:
My strategy was to check that the mode string (ex `u=rw`) containes a `=` anywhere and no `+/-` which is the bare minimum check to say it is an assignment.

purpleidea replied:
Oh, okay. I'm okay with this, can you please add a short, clear sentence comment explaining that logic at the top of the `if` please? I might just be out of it today, but since this is sort of untraditional for parsing style, a quick reminder will help! =D

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+		who, what = sp[0], sp[1]
+	}
+
+	// Validate input
```
Short comments are either lower case snippets or full sentences that are capitalized and end with a period. Same comment for other places. https://github.com/purpleidea/mgmt/blob/master/docs/style-guide.md#comments

purpleidea commented on util/sym-mode_test.go:
```diff
@@ -0,0 +1,58 @@
+	for _, test := range symModeTests {
+		t.Run(test.input, func(t *testing.T) {
+			if got, err := util.SymbolicMode(test.input, def); err != nil {
+				t.Fatalf("Failed to parse the symbolic string: %v", err)
```
Errorf instead of fatal please. You can return or continue as needed.

purpleidea commented on util/sym-mode_test.go:
```diff
@@ -0,0 +1,58 @@
+	def := os.FileMode(0644) | os.ModeSetgid
+	symModeTests := []struct {
+		input  string
+		expect os.FileMode
```
Please look at our other tests to add an `experr` sort of variable or similar. The recent pr by @dbakong shows this too.

dgbuckley replied:
In that PR the errors are being compared and even though  the text is the same they are separate error values so it won't pass that way ([playground](https://play.golang.org/p/nEYvEGxgVg3)). Would it be preferred to check for nil then compare output of `Error()` or could error constants and `errors.Is` be used?

purpleidea replied:
If we made that error compare somewhere, please let me know! In general it's best to compare error constants, or if it's part of a test case and we're checking for a particular error that the user never needs to compare to, then the string compare is okay.

purpleidea commented on util/sym-mode_test.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	for _, test := range symModeTests {
+		t.Run(test.input, func(t *testing.T) {
```
Needs to be race free... See other tests to avoid this. https://golang.org/doc/articles/race_detector.html#Race_on_loop_counter

dgbuckley replied:
Sorry I thought range with `t.Run` was race free because each run has its own struct from the slice. Does a copy need to be made before `t.Run()`?

purpleidea replied:
Honestly I have to dig into the code again and check, but at one time I'm fairly sure it wasn't race free. So I did this: https://github.com/purpleidea/mgmt/blob/2ea492c96512a0802c98c9a06984457e88d5f2a4/lang/interpret_test.go#L431

purpleidea commented on util/sym-mode.go:
```diff
@@ -0,0 +1,262 @@
+// Mgmt
```
Please move this file to engine/util/ and name it mode.go or something without a dash in it please. (Same with test.)

purpleidea commented:
On Thu, Oct 17, 2019 at 4:01 PM dgbuckley <notifications@github.com> wrote:
>
> @purpleidea that sounds great. No rush on this PR, I'm just excited to be contributing to this awesome project. Also sorry about the overlap of work; I'll do better to watch PRs moving forward.

It's not a problem. We're happy to have you as a new contributor. I
just finished the review. Please feel free to join us in #mgmtconfig
on Freenode IRC or ping me wherever if you're looking for more hacking
with us! Happy to help you find something to work on if you don't
already have something in mind.

dgbuckley commented:
@purpleidea I think I covered all of your feedback. I changed the main function to accept a slice of modes and fail safely if there is repeated subjects. I also added the support for your suggestion to make +/- toggle with a bool and the missing 'a' subject I realized I forgot.

purpleidea commented on engine/resources/file.go:
```diff
@@ -137,10 +137,25 @@ func (obj *FileRes) isDir() bool {
+	}
+
+	// Try parsing symbolic by first getting the files current mode.
+	m := os.FileMode(0)
```
You can inline this section and not do this awkward initialization. Eg:

```
stat, err := ...
if err != nil {
...
}
m := stat.Mode()
```

Better?

purpleidea commented on engine/resources/file.go:
```diff
@@ -137,10 +137,25 @@ func (obj *FileRes) isDir() bool {
+	}
+
+	modes := strings.Split(obj.Mode, ",")
+	var err error
```
and then you can nuke this line

purpleidea commented on engine/util/mode.go:
```diff
@@ -0,0 +1,297 @@
+
+type ModeError string
+
+func (m ModeError) Error() string { return string("invalid mode: " + m) }
```
fmt.Sprintf instead please.

purpleidea commented on engine/util/mode.go:
```diff
@@ -0,0 +1,297 @@
+
+type ModeError string
+
+func (m ModeError) Error() string { return string("invalid mode: " + m) }
```
https://github.com/purpleidea/mgmt/blob/master/docs/style-guide.md#method-receiver-naming please.

purpleidea replied:
Note: I'm not actually sure why you're defining a special error type unless you want to check against it somewhere. Eg:

https://github.com/purpleidea/mgmt/blob/06952c224b734b5146539d1fedc953e7b1a50041/lang/interfaces/error.go#L29

If you're planning to, go ahead, if not you can just use: `fmt.Errorf` in the error returns.

purpleidea commented on engine/util/mode.go:
```diff
@@ -0,0 +1,297 @@
+	for i, _ := range symModes {
+		if strings.ContainsRune(symModes[i].who, 'a') || symModes[i].who == "" {
+			// If 'a' or empty who (implicit 'a') is called and there are more
+			// symbolic modes in the slice then it must be a repitition.
```
repetition repetition

purpleidea commented:
PS: This commit should also remove this line and update the docstring above it:

https://github.com/purpleidea/mgmt/blob/06952c224b734b5146539d1fedc953e7b1a50041/engine/resources/file.go#L98

And the docs/resources.md file mode section too please!

dgbuckley commented:
@purpleidea. I think I've covered all of your feedback. I removed the error type because I only added it for testing and the same can be achieved with checking the error string.

purpleidea commented:
@dgbuckley Doesn't build: https://travis-ci.org/purpleidea/mgmt/jobs/602558300#L837 and sorry commit will fail unless it's `engine: resources: Add symbolic mode to file resource` (the resources prefix needs an S)

purpleidea commented:
@dgbuckley I'm planning on merging this shortly, thanks! There are a few nitpicks that I was planning on adding in a subsequent commit. Is this okay with you or would you prefer I enumerate them for you to fix up instead? I only mention because they're possibly easier to understand as a diff.

Thanks again!

dgbuckley commented:
You can definitely make any changes in a later commit; I look forward to seeing the diff. Thank you for the feedback you have provides.

purpleidea commented:
This patch has now been rebased to master and merged! Thank you again!
As discussed I made a few cleanups to get this in. This is in 1eba5833d5418b9473b9e961a455c7ae2e9466fe I've also updated the style guide so that I can better provide feedback about things to change. Please have a look at that commit, and let me know if you agree with it all or if you have any questions!

Thank you again for your patch, and hopefully we can persuade you to write something else! -- Adding a Reverse method to an existing resource might be a good one!

Cheers!
